### ISSUE TITLE
declare_interior_mutable_const triggers on array initializer

### ISSUE THREAD

**Issue by quaternic** [rust-lang/rust-clippy#7665]
Lint name:
declare_interior_mutable_const

I tried this code:
```rust
let _: [Cell<bool>; 7] = [Cell::new(true); 7];
```
Since the array initialization syntax requires `T: Copy`, rustc errors and suggests this fix that works correctly:
```rust
const TRUE_CELL: Cell<bool> = Cell::new(true);
let _: [Cell<bool>; 7] = [TRUE_CELL; 7];
```
[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=bd791ba365616f1409a8fd510f98dcce)

To my knowledge this is the best way to initialize an array with a `const` but `!Copy` value. 
([stackoverflow answer](https://stackoverflow.com/a/66776497))

However, `declare_interior_mutable_const` triggers and incorrectly suggests "make this a static item (maybe with lazy_static)", which doesn't actually work at all. The lint documentation claims the code is bad because "Consts are copied everywhere they are referenced, ... , which defeats the whole purpose of using these types in the first place.", but in this case that is the entire reason for using a `const` item.

The known issues mention "a legacy way to supply an initialized value to downstream static items" as a legit use, but it is not immediately clear if this case is covered by that.

**Comment by xFrednet** [comment]
Hey, thank you for the report :upside_down_face:. This is actually a difficult case, where this way of initializing the array uses the fact that each `const` value is copied. This is copying is the exact thing that the lint, tries to warn for as is can be surprising to newcomers if this is not intended. 

For cases like this, I would suggest adding a `#[allow(clippy::declare_interior_mutable_const)]` attribute to the `const` value as you intentionally use this behavior. Currently, I can sadly not think of a simple way to fix this, as we would need to check that the `const` value is only used for array initialization. It is possible to check for the usage of a value, but we would need to check the entire crate for global constant values. A good compromise could be to only check if a `const` value is used to initialize an array if the code is inside a function body. That would limit the scope of the search.

As this is theoretically working as intended, I'll mark this as a possible enhancement of the lint :upside_down_face:

@rustbot label -C-Bug -I-false-positive +C-enhancement

**Comment by quaternic** [comment]
> This is copying is the exact thing that the lint, tries to warn for as is can be surprising to newcomers if this is not intended.

Wouldn't that be just as surprising if the `const` is not interior-mutable?

The example in the documentation *also* triggers [borrow_interior_mutable_const](https://rust-lang.github.io/rust-clippy/master/index.html#borrow_interior_mutable_const), so it's not clear why *this* lint is needed in addition. In fact, it seems to me that any attempt to mutate the `const` requires a shared reference and triggers borrow_interior_mutable_const. If there is no such attempt, the interior mutability is inconsequential.

**Comment by xFrednet** [comment]
> > This is copying is the exact thing that the lint, tries to warn for as is can be surprising to newcomers if this is not intended.
> 
> Wouldn't that be just as surprising if the `const` is not interior-mutable?

It shouldn't, because you can't modify constant values. Here is an example, what the lint is actually about 
[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=583c2cf8bce66a5cbf009452c7319c43):

```rs
const GLOBAL_CELL: Cell<bool> = Cell::new(false);
fn main() {
    GLOBAL_CELL.set(true);
    println!("{:?}", GLOBAL_CELL)
}
```


This will print out "`Cell { value: false }`" which might be counterintuitive. The lint basically tries to warn if a user tries to use `const` to create global values, as they are not truly global and places inside the code

> The example in the documentation _also_ triggers [borrow_interior_mutable_const](https://rust-lang.github.io/rust-clippy/master/index.html#borrow_interior_mutable_const), so it's not clear why _this_ lint is needed in addition. In fact, it seems to me that any attempt to mutate the `const` requires a shared reference and triggers borrow_interior_mutable_const. If there is no such attempt, the interior mutability is inconsequential.

You're right, those lints have some overlaps. I'm not sure why, my guess is that `declare_interior_mutable_const` is for constant values in general and `borrow_interior_mutable_const` is an additional lint in case you use such a `const` value from an external crate. But I'm not sure on this one.

**Comment by hniksic** [comment]
I stumbled upon this issue. It was in code like this:

```rust
static REGISTERED_NAMES: [OnceLock<String>; 256] = {
    const INIT_HELPER: OnceLock<String> = OnceLock::new();
    [INIT_HELPER; 256]
};
```

This `const` is the only way to initialize the static that I'm aware of, advertised e.g. in answers to [this StackOverflow question](https://stackoverflow.com/questions/28656387/initialize-a-large-fixed-size-array-with-non-copy-types).

**Comment by Jarcho** [comment]
As of 1.79 you can use inline const blocks (e.g. `[const { Cell::new(0) }; 5]`).
